fix(spec/consensus): Small fixes to overview.md file.

nenadmilosevic95 opened the pull request:
I made a pass on overview.md file. 
IMHO this is a very nice explanation of Tendermint. Great job @cason and @josef-widder and who else contributed !!! 
I made some minor fixes, and made some comments. I am available if you want to discuss anything, and feel free to ignore the things you think are not relevant! 

### PR author checklist

#### For all contributors

- [ ] Reference GitHub issue
- [x] Ensure PR title follows the [conventional commits][conv-commits] spec

#### For external contributors

- [x] Maintainers of Malachite are [allowed to push changes to the branch][gh-edit-branch]

[conv-commits]: https://www.conventionalcommits.org/en/v1.0.0/#summary
[gh-edit-branch]: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests

cason reviewed (changes requested):
Thank you very much for taking your time to review this.

I tried to fix some of the points you have raised and we need to further discuss others.

The point is that we can't commit your `\NM{}` comments to `main`, so either we keep it as a draft, or we work to rephrase some content in order to address the comments.

cason commented on specs/consensus/overview.md:
```diff
@@ -984,14 +996,14 @@ such a good round. Here, there are two crucial points:
 
 1. The [`proposer(h, r)` function](#proposer-selection) returns a correct process `p`;
-2. And `validRound_p` equals the maximum `validRound_q` among every correct process `q`.
+2. And `validRound_p` equals the maximum `lockedRound_q` among every correct process `q`. \NM{It can be also larger, so >=}
```
```suggestion
2. And `validRound_p` is larger or equals the maximum `lockedRound_q` among every correct process `q`.
```

cason commented on specs/consensus/overview.md:
```diff
@@ -1047,14 +1063,14 @@ successful round is observed.
-This is achieved by the means of, and it is actually the reason for which
-Tendermint has, the `timeoutPrecommit(r*)`.
+This is achieved by the means of, and it is actually the key reason for which
+Tendermint has, the `timeoutPrecommit(r*)`. 
```
```suggestion
Tendermint has, the `timeoutPrecommit(r*)`.
```

cason commented on specs/consensus/overview.md:
```diff
@@ -1047,14 +1063,14 @@ successful round is observed.
 As a result, if a correct process has locked or updated its valid value in
-round `r*`, then every correct process will at least update its valid value in
-round `r*`.
+round `r*`,  then every correct process will at least update its valid value in
```
```suggestion
round `r*`, then every correct process will at least update its valid value in
```

cason commented on specs/consensus/overview.md:
```diff
@@ -1047,14 +1063,14 @@ successful round is observed.
-round `r*`, then every correct process will at least update its valid value in
-round `r*`.
+round `r*`,  then every correct process will at least update its valid value in
+round `r*`, because due to gossip communication property it will receive the same set of messages. 
```
```suggestion
round `r*`.
This happens because, due to [Gossip communication property](#network), every correct process receives the same set of messages within `∆`. 
```

cason commented on specs/consensus/overview.md:
```diff
@@ -1019,6 +1031,10 @@ Notice, however, that from the [Gossip communication property](#network), `q`
 should eventually receive the POL for `v` in round `vr`, since `p` is a correct
 process.
 
+\NM{I think we should mention in the text here that we show the intution how condition 2 is achieved later, because
```
I agree. Added a sentence to the last paragraph, check if it works. :)

cason commented on specs/consensus/overview.md:
```diff
@@ -1019,6 +1031,10 @@ Notice, however, that from the [Gossip communication property](#network), `q`
 should eventually receive the POL for `v` in round `vr`, since `p` is a correct
 process.
 
```
```suggestion
And, as we detail on the next section, when the synchrony assumptions are observed, `p` will receive such messages before it receives the new `PROPOSAL`.
```

cason reviewed (commented):
I have now completed covering your reviews.

Can you check the suggestions that are good, or provide alternative version of them, so we can remove some of the added "comments"?

josef-widder commented on specs/consensus/overview.md:
```diff
@@ -39,7 +39,7 @@ A this point, the process increases `h_p` (line 52) and starts the next height
 
 For the sake of the operation of the consensus algorithm, heights are
-completely independent executions.
+completely independent executions. NM{Not sure I fully understand what do we want to say here?}
```
```suggestion
communication-closed, that is, a message from some height `H` does not influence a process that is in height `H' != H`.
```

josef-widder commented on specs/consensus/overview.md:
```diff
@@ -180,7 +180,8 @@ associated to a [round step](#round-steps):
   value `v` for which the process has received `⟨PREVOTE, h, r, id(v)⟩`
-  messages from a super-majority of processes, or the special `nil` value otherwise.
+  messages from a [super-majority](#voting-power) of processes, or the special `nil` value otherwise.
+  NM{I am not sure we have defined super-majority anywhere?} 
```
```suggestion
```
I think @cason already addressed this

josef-widder commented on specs/consensus/overview.md:
```diff
@@ -368,12 +372,12 @@ inducing undesirable behaviour, are two:
 Different (equivocating) versions of the same message from the same sender
 should, from a defensive point of view, be disregarded and dropped by the
-consensus logic as they were duplicated messages.
+consensus logic as they were duplicated messages. NM{Ok I guess this is the change you want to implement or have alrady implemented, that is different from the paper?}
```
```suggestion
consensus logic as they were duplicated messages (note that doing this maintains safety, but leads to violations of the gossip property from the [Tendermint paper](https://arxiv.org/abs/1807.04938), and thus introduces liveness issues; see below). 
```

josef-widder commented on specs/consensus/overview.md:
```diff
@@ -744,7 +748,7 @@ But `∆` is not always observed by the system, which may operate asynchronously
 There is, however, a (possibly unknown) Global Stabilization Time (`GST`), a
 time from which the system becomes synchronous and `∆` is observed for all
-messages sent by correct processes.
+messages sent by correct processes. \NM{Ok this "possibly unknown" in both sentences is confusing me? Did you try to somehow capture both versions of partially synchronous system models? }
```
I think it is fine, I would just keep it. Even if both are unknown, the algorithm should work, no? That would be the model in the Chandra/Toueg paper on failure detectors.

josef-widder commented on specs/consensus/overview.md:
```diff
@@ -907,6 +913,11 @@ value in round `r`.
 This is the reason for which pseudo-code lines 15-16 adopt `validValue_p`
 instead of `lockedValue_p`.
 
+\NM{The explanations here are fine for me, but I am not good to review this as I know in detail how this mechanism 
+works. I think we should ask someone who does not understand how valid value and locked value correlate and see
+if he/she got it from this. I think this explanation introduces the difference but do not explain the mechanism. 
+Update: I saw that latter you do explain, so I am happy with the explanations. }
```
```suggestion
```
I guess the update says that it is fine as is.

josef-widder reviewed (commented):
Thanks for the improvements. I left a few suggestions.

nenadmilosevic95 commented:
@josef-widder I still did not have time to go through your comments. Let me do that first and then we can schedule the call if still necessary!

nenadmilosevic95 commented on specs/consensus/overview.md:
```diff
@@ -838,12 +842,12 @@ unaware of the lock in round `r`, therefore propose new values.
+possible to produce another lock in a round `r' > r`. \NM{These processes do not need to be Byzantine.}
 None of the locked values can be decided, for the lack of votes, but liveness
-is under threat, as detailed in this [discussion][equivocation-discussion].
+is under threat, as detailed in this [discussion][equivocation-discussion]. \NM{Potentially they cannot be decided, so maybe we can rephrase a bit this sentence.  }
```
This one we can briefly discuss

josef-widder replied:
@nenadmilosevic95 takes a stab at this paragraph.

nenadmilosevic95 commented on specs/consensus/overview.md:
```diff
@@ -879,6 +883,8 @@ There are however some scenarios to consider:
 3. `step_p = precommit`: in this case, `p` only updates `validValue_p` to `v`
    and `validRound_p` to `r`.
 
+\NM{I would maybe just talk about two scenarios here, when it locks and when it just updates valid values}
```
Do you want me to make this change or you want to keep it like this?

josef-widder replied:
Yes. let's make two cases.

nenadmilosevic95 commented:
@josef-widder  @cason I went through your comments, let's have a short call when you have time to discuss few things and close this PR.

josef-widder commented on specs/consensus/overview.md:
```diff
@@ -257,20 +257,23 @@ be a priori known by all consensus processes.
 So, a priori, the algorithm expects a correct process to store all the
 multiple proposals broadcast by a Byzantine proposer.
 Which, by itself, constitutes an attack vector to be considered.
+NM{I don't know how this is implemented in the code, but in a round process do need to store 
+all values but if one of the values collects 2f+1 prevotes it can discard others? }
```
```suggestion
```

josef-widder commented on specs/consensus/overview.md:
```diff
@@ -368,12 +372,12 @@ inducing undesirable behaviour, are two:
 Different (equivocating) versions of the same message from the same sender
 should, from a defensive point of view, be disregarded and dropped by the
-consensus logic as they were duplicated messages.
+consensus logic as they were duplicated messages. NM{Ok I guess this is the change you want to implement or have alrady implemented, that is different from the paper?}
```
These two paragraphs are more related to gossip. Let's move them to the end of the "Network" section

josef-widder commented on specs/consensus/overview.md:
```diff
@@ -744,7 +747,7 @@ But `∆` is not always observed by the system, which may operate asynchronously
 There is, however, a (possibly unknown) Global Stabilization Time (`GST`), a
 time from which the system becomes synchronous and `∆` is observed for all
-messages sent by correct processes.
+messages sent by correct processes. \NM{Ok this "possibly unknown" in both sentences is confusing me? Did you try to somehow capture both versions of partially synchronous system models? }
```
```suggestion
messages sent by correct processes.
```

josef-widder commented on specs/consensus/overview.md:
```diff
@@ -815,7 +818,7 @@ The rationale is that `p` _could have_ locked and issued a `PRECOMMIT` for `v`
 round step of round `vr`.
 If `p` could have locked `v` in round `vr`, then any correct process could
-have produced a valid lock in round `vr`.
+have produced a valid lock in round `vr`. \NM{Ok, this explanation is too high level for me but I understand that it is hard to explain at this point that it can accept it because it knows that lockedValue is not decided}
```
```suggestion
have produced a valid lock in round `vr`. 
```

josef-widder commented on specs/consensus/overview.md:
```diff
@@ -838,12 +842,12 @@ unaware of the lock in round `r`, therefore propose new values.
 accepted by correct processes that have no locks.
 With the votes of Byzantine processes, which may misbehave, it is then
-possible to produce another lock in a round `r' > r`.
+possible to produce another lock in a round `r' > r`. \NM{These processes do not need to be Byzantine.}
```
Let's fix the sentence to:
It is then possible to produce another lock in a round `r' > r`.

cason replied:
This is a liveness discussion. I would rather remove this paragraph as the reason for which we might change locks is explained in the new version of the previous paragraphs.

josef-widder commented on specs/consensus/overview.md:
```diff
@@ -929,7 +940,8 @@ correct processes, and consider the [Locked Value](#locked-value) handling:
 - every process `p` in `C` will broadcast a `PREVOTE` for `nil` in rounds
-  `r' > r` where a value `v' != v` is proposed.
+  `r' > r` where a value `v' != v` is proposed. \NM{this might not be clear for all and this is 
+  what we explain below, so maybe we can somehow rephrase" }
```
let's clarify assumption and consequences here (first two bullets should be in an "if", and the third should be the consequence.

cason commented on specs/consensus/overview.md:
```diff
@@ -368,12 +368,12 @@ inducing undesirable behaviour, are two:
 Different (equivocating) versions of the same message from the same sender
 should, from a defensive point of view, be disregarded and dropped by the
-consensus logic as they were duplicated messages.
+consensus logic as they were duplicated messages. 
```
```suggestion
consensus logic as they were duplicated messages.
```

cason commented on specs/consensus/overview.md:
```diff
@@ -827,20 +832,21 @@ And more recent (from high-numbered rounds) locks prevail.
 > But this would require more than one third of the voting power to be owned by
 > Byzantine processes, which is a violation of Tendermint's failure model.
 
+\NM{I think with the current version of this paragraph the text below is not necessary.}
```
I agree. We can remove all the content until line 852.

cason commented on specs/consensus/overview.md:
```diff
@@ -899,14 +904,15 @@ eventually expired (lines 61-64), leading `p` to broadcast a `PRECOMMIT` for `ni
 This is the reason for which pseudo-code lines 15-16 adopt `validValue_p`
 instead of `lockedValue_p`.
 
+
```
```suggestion
```

cason reviewed (approved):
Thank for this.

I think we have only a paragraph in the Locked Value section to remove and we are good to merge.

nenadmilosevic95 commented:
@josef-widder @cason  I made final pass. I removed the paragraph and added two sentences to valid value section. The sections with Byzantine voters and proposers are still where they were, I guess this is fine for now.

cason commented:
If @josef-widder does not have any additional concern, you can merge it :)
